### ISSUE TITLE
Failed authentication not throwing Throttling Exception

### ISSUE THREAD

**Issue by benwrigley** [cartalyst/sentinel#451]
I've replicated this in tinker and phpunit so far.

I have a user throttling set to 5 in my config.

using tinker if I run:

`Sentinel::authenticate(['email' => 'real@user.email', 'password' => 'wrong-password'])`

I get `false` as expected.

keep running the same command I keep getting `false` no matter hoe many times, the `ThrottlingException` is never thrown.

Unless I restart tinker and then run the command again. Then I get the exception.

Any thoughts as this is also affecting my feature tests.

**Comment by talal424** [comment]
did you wrap it with try catch ?

**Comment by benwrigley** [comment]
@talal424 

no, this issue is not how to catch it, but that the Exception simply isn't thrown when it should be.

**Comment by brunogaspar** [comment]
@benwrigley Care to share one of the affected feature tests?

Also, is it reproducible outside of tests, like, real world usage?

**Comment by benwrigley** [comment]
hi @brunogaspar 

No I can't replicate in the real world and the issue doesn't occur in dusk tests (I'm guessing this might be because the session is renewed each step?) 

Here is the test:

```
    public function test_wrong_login_locks_account()
    {

        $loginCredentials = [
            'email' => 'real@user.account',
            'password' => 'wrong_password'
        ];

        $loginAttempts=5;

        foreach (range(0, $loginAttempts) as $i) {
            $this->fromUrl(route('login'))
            ->post('/login', $loginCredentials)
            ->assertEIMErrorExists('login_fail')
            ->followRedirects();
        }

        $this->fromUrl(route('login'))
        ->post('/login', $loginCredentials)
        ->assertEIMErrorExists('login_locked');
    }
```

**Comment by talal424** [comment]
@benwrigley 
yeah it was a dumb comment :laughing:

**Comment by brunogaspar** [comment]
@benwrigley Just did a quick test with tinker and in fact i cannot see the Exception, however, if i close the tinker session and reopen it then run the authenticate command, it will show the exception.

So it seems a thing with tinker itself, not sure if it's a bug within Sentinel though.

Going to close, but feel free to reopen if you still find it's something within Sentinel we can tweak.

**Comment by benwrigley** [comment]
Hi @brunogaspar 

I don't think it's just a tinker issue as I am having the problem in my feature tests too so I think it's still worth investigating?

Thanks

**Comment by brunogaspar** [comment]
I'll reopen, but i don't have enough time today to revisit this though.

**Comment by benwrigley** [comment]
@brunogaspar of course, totally understand. Thank you.

**Comment by 8633brown** [comment]
looking into this a little bit i've created somewhat of an integration test to help in testing this. [here](https://github.com/8633brown/sentinel/blob/throttlingbug/tests/throttlingbugTest.php)
(this needs #487 to run correctly)

i think i've narrowed the bug down to these lines
https://github.com/cartalyst/sentinel/blob/d1e5d6c2afab33134c70eac1962aed769ddc3077/src/Throttling/IlluminateThrottleRepository.php#L456-L458
i think what is happening is when using ```Native\Facades\Sentinel.php``` it's creating the static instance for sentinel and it is logging every login attempt to the database correctly. however when reading these back from the database the first time we assign them to the variable ```$this->userThrottles``` now we have assigned the user throttles as an array with the user id as the key. the next time we login and attempt to read the user throttles again we find the variable we have just written to which previously was written as 0. and we never run the line ```$this->userThrottles[$key] = $this->loadUserThrottles($user);``` to retrieve the new values from the database.

**Comment by 8633brown** [comment]
TLDR; i think reading and writing throttles to the DB need to be consistent with each other where as at the moment they are not. 

This bug is actually the same for global and ip intervals too. 
https://github.com/cartalyst/sentinel/blob/d1e5d6c2afab33134c70eac1962aed769ddc3077/src/Throttling/IlluminateThrottleRepository.php#L419-L421
https://github.com/cartalyst/sentinel/blob/d1e5d6c2afab33134c70eac1962aed769ddc3077/src/Throttling/IlluminateThrottleRepository.php#L387-L389
I assume this was to stop reading from the DB on every call to ```Sentinel::authenticate```?

I think theres two options here. either remove these lines
https://github.com/cartalyst/sentinel/blob/d1e5d6c2afab33134c70eac1962aed769ddc3077/src/Throttling/IlluminateThrottleRepository.php#L387
https://github.com/cartalyst/sentinel/blob/d1e5d6c2afab33134c70eac1962aed769ddc3077/src/Throttling/IlluminateThrottleRepository.php#L419
https://github.com/cartalyst/sentinel/blob/d1e5d6c2afab33134c70eac1962aed769ddc3077/src/Throttling/IlluminateThrottleRepository.php#L456
so that the throttles are checked on every call. the same as writing throttles to the DB are. This would fix the bug and make testing possible too.

or when writing throttles to the DB, the same arrays need to be checked to ensure we only write to the DB once. This would still fix the bug however testing throttling throws an exception externally would be impossible right now by using ```Sentinel::authenticate```.

I'm unsure as to which would be the best fix.

**Comment by suwardany** [comment]
@benwrigley i've pushed a little tweak there that should address the issue .. give it a try and let me know how it's looking on your end!

@8633brown thanks for digging in here .. got any thoughts on that fix?

**Comment by 8633brown** [comment]
I've been trying to understand if there could ever be a situation for logging multiple users in on the same request. 
In this situation is it not possible for the ip and user throttles to be stuck on the original user login attempt? 
Would it be a better fix to only read and write to the DB when the ip address or the user is different from the previous attempt?
